lib/scholar/linear/isotonic_regression.ex

```diff
@@ -0,0 +1,419 @@
+          [1.0, 3.0, 6.0, 8.0, 9.0, 10.0]
+        ),
+        increasing?: true,
+        sample_weights: Nx.tensor(
+          [1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
+        ),
+        cutoff_index: Nx.tensor(
+          5
+        )
+      }
+  """
+  deftransform fit(x, y, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    opts =
+      [
+        sample_weights_flag: opts[:sample_weights] != nil
+      ] ++
+        opts
+
+    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, 1.0)
+    x_type = to_float_type(x)
+    x = to_float(x)
+    y = to_float(y)
+
+    sample_weights =
+      if Nx.is_tensor(sample_weights),
+        do: Nx.as_type(sample_weights, x_type),
+        else: Nx.tensor(sample_weights, type: x_type)
+
+    sample_weights = Nx.broadcast(sample_weights, {Nx.axis_size(y, 0)})
+
+    fit_n(x, y, sample_weights, opts)
+  end
+
+  defnp fit_n(x, y, sample_weights, opts) do
+    {x_min, x_max, x_unique, y, index_cut} = build_y(x, y, sample_weights, opts)
+
+    %__MODULE__{
+      x_min: x_min,
+      x_max: x_max,
+      x_thresholds: x_unique,
+      y_thresholds: y,
+      increasing?: opts[:increasing?],
+      cutoff_index: index_cut
+    }
+  end
+
+  @doc """
+  Makes predictions with the given `model` on input `x` and interpolating `function`.
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 4, 7, 9, 10, 11])
+      iex> y = Nx.tensor([1, 3, 6, 8, 9, 10])
+      iex> model = Scholar.Linear.IsotonicRegression.fit(x, y)
+      iex> function = Scholar.Linear.IsotonicRegression.build_function(model)
+      iex> to_predict = Nx.tensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
+      iex> Scholar.Linear.IsotonicRegression.predict(function, model, to_predict)
+      #Nx.Tensor<
+        f32[10]
+        [1.0, 1.6666667461395264, 2.3333334922790527, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
+      >
+  """
+  defn predict(function, model, x) do
+    check_input_shape(x)
+    x = Nx.flatten(x)
+    x = Nx.clip(x, model.x_min, model.x_max)
+    Scholar.Interpolation.Linear.predict(function, x)
+  end
+
+  @doc """
+  Builds a function for predictions with the given `model`.
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 4, 7, 9, 10, 11])
+      iex> y = Nx.tensor([1, 3, 6, 8, 9, 10])
+      iex> model = Scholar.Linear.IsotonicRegression.fit(x, y)
+      iex> function = Scholar.Linear.IsotonicRegression.build_function(model)
+      %Scholar.Interpolation.Linear{
+        coefficients: Nx.tensor(
+          [
+            [0.6666666865348816, 0.3333333134651184],
+            [1.0, -1.0],
+            [1.0, -1.0],
+            [1.0, -1.0],
+            [1.0, -1.0]
+          ]
+        ),
+        x: Nx.tensor(
+          [1.0, 4.0, 7.0, 9.0, 10.0]
+        )
+      }
+  """
+  def build_function(model) do
+    cutoff = Nx.to_number(model.cutoff_index)
+    x = model.x_thresholds[0..cutoff]
+    y = model.y_thresholds[0..cutoff]
+
+    if Nx.size(y) == 1 do
+      Nx.broadcast(y, Nx.shape(x))
```

Comment:
Will this clause work? Can we pass a tensor as a function to `Scholar.Interpolation.Linear.predict`?

Reply:
Good catch, now it won't work. It should be called instead of interpolation `predict` in `Scholar.Linear.Isotonic.predict`